Gems/Atom/RHI/Code/Source/RHI/MultiDeviceSwapChain.cpp

```diff
@@ -0,0 +1,278 @@
+    {
+        IterateObjects<SwapChain>([&verticalSyncInterval]([[maybe_unused]] auto deviceIndex, auto deviceSwapChain)
+        {
+            deviceSwapChain->SetVerticalSyncInterval(verticalSyncInterval);
+        });
+
+        m_descriptor.m_verticalSyncInterval = verticalSyncInterval;
+    }
+
+    const AttachmentId& MultiDeviceSwapChain::GetAttachmentId() const
+    {
+        return m_descriptor.m_attachmentId;
+    }
+
+    const SwapChainDescriptor& MultiDeviceSwapChain::GetDescriptor() const
+    {
+        return m_descriptor;
+    }
+
+    bool MultiDeviceSwapChain::IsExclusiveFullScreenPreferred() const
+    {
+        auto result{ true };
+
+        IterateObjects<SwapChain>([&result]([[maybe_unused]] auto deviceIndex, auto deviceSwapChain)
+        {
+            result |= deviceSwapChain->IsExclusiveFullScreenPreferred();
+        });
+
+        return result;
+    }
+
+    bool MultiDeviceSwapChain::GetExclusiveFullScreenState() const
+    {
+        auto result{ true };
+
+        IterateObjects<SwapChain>([&result]([[maybe_unused]] auto deviceIndex, auto deviceSwapChain)
+        {
+            result &= deviceSwapChain->GetExclusiveFullScreenState();
+        });
+
+        return result;
+    }
+
+    bool MultiDeviceSwapChain::SetExclusiveFullScreenState(bool fullScreenState)
+    {
+        auto result{ true };
+
+        IterateObjects<SwapChain>([&result, fullScreenState]([[maybe_unused]] auto deviceIndex, auto deviceSwapChain)
+        {
+            result &= deviceSwapChain->SetExclusiveFullScreenState(fullScreenState);
+        });
+
+        return result;
+    }
+
+    void MultiDeviceSwapChain::ProcessRecreation()
+    {
+        IterateObjects<SwapChain>([]([[maybe_unused]] auto deviceIndex, auto deviceSwapChain)
+        {
+            deviceSwapChain->ProcessRecreation();
+        });
+    }
+
+    uint32_t MultiDeviceSwapChain::GetImageCount() const
+    {
+        return aznumeric_cast<uint32_t>(m_mdImages.size());
+    }
+
+    MultiDeviceImage* MultiDeviceSwapChain::GetCurrentImage() const
+    {
+        if (m_descriptor.m_isXrSwapChain)
+        {
+            return m_mdImages[m_xrSystem->GetCurrentImageIndex(m_descriptor.m_xrSwapChainIndex)].get();
+        }
+        AZ_Error("Swapchain", !m_deviceObjects.empty(), "No device swapchain image available.");
+        // Note: Taking the current swapchain image index from the first device swap chain if there are multiple
+        auto currentImageIndex{ AZStd::static_pointer_cast<SwapChain>(m_deviceObjects.begin()->second)->GetCurrentImageIndex() };
+        return m_mdImages[currentImageIndex].get();
+    }
+
+    MultiDeviceImage* MultiDeviceSwapChain::GetImage(uint32_t index) const
+    {
+        return m_mdImages[index].get();
+    }
+
+    void MultiDeviceSwapChain::Present()
+    {
+        IterateObjects<SwapChain>([]([[maybe_unused]] auto deviceIndex, auto deviceSwapChain)
+        {
+            deviceSwapChain->Present();
+        });
+    }
```

Comment:
Well no, we haven't completely figured this out yet, because that's something we cannot do without looking at how the RPI specific changes will work exactly. So at this point we simply made everything uniformly happen on all devices, though I agree it doesn't make too much sense here. Could turn out that swap chains will only ever be used with a single device in the end, then this doesn't matter, but we simply can't tell at the moment. So this may be adapted later. For now I don't think we can predict yet, what an ideal solution would be here.

Reply:
I agree with Akio here. Adding MDSwapchain doesn't make sense and even some of these functions in this class doesn't make sense. For example the code around handling XR related swap chain images or dealing with ExclusiveFullScreen related code. I think at this time we should remove this class and add it back when we have a better picture of why and how it will be used.

Reply:
At some point if we decide to build a specific MGPU pipeline where some sections of swap chain texture is built by different gpu we could consider a MDswapchain which would be responsible for copying/synchronizing the result into one texture. Even then a MDswapchain functionality will look very different and we can consider adding this class at that time.

Reply:
You're absolutely right that it doesn't make sense, the way it is implemented at the moment if a `MultiDeviceSwapChain` would be instantiated with a device mask with multiple bits set.

Unfortunately, we cannot get rid of the class completely, we still need to get `MultiDeviceImage`s out of the swap chain to later work with all the changes in the RPI. Therefore, I changed the class to accept only a device index rather than a mask to make sure that only a single bit is then set in the device mask.

This is how we have been using the `MultiDeviceSwapChain` in our RPI changes anyway, so we never even tried setting multiple bits in the device mask.

Reply:
Are you able to provide context into where RPI needs MultiDeviceImages from swap chain? Maybe I am missing something but it shouldn't need to but let's assume it does need it. In that case I think it is still better that if we just modify RHI::Swapchain's RHI::Image to RHI::MDImage as part of that change. It would prevent adding a duplicate class unnecessarily.

Reply:
I've removed the `MultiDeviceSwapChain` from this PR to progress with the more important parts and will create a new PR with just it once this one has been merged to continue the discussion there. The place where the MDImage of a SwapChain is needed is when it is used in an attachment. We could avoid them altogether (see https://github.com/o3de/sig-graphics-audio/issues/120#issuecomment-1674386664) if we would make the transition to single device resources earlier than between frame graph compile and execute, i.e., before compile already, as discussed in the linked comment. However, I have dropped that in the meantime idea since it is probably less flexible if we want to do some form of automated scheduling (e.g. figuring out in the frame graph when a resource needs to be copied from one device to another). Therefore, we need MD resources for attachments, since those are handled during compilation already where we are still MD and `SwapChainFrameAttachment` clearly shows, where the MDSwapChain would be necessary. I'm not a big fan of having the MDImages in the SDSwapChain class. This would be possible, but pretty inconsistent with the rest of the system. Suddenly all backend SwapChain implementations would have to consider those to some extent, since they are deriving from the SDSwapChain base class and not having to bother about MD (not 100 % possible but as much as possible) in the backends is one of the things we wanted to achieve with these MD classes.

Reply:
Thanks for making the change. I have approved and started an AR run.

I see your point about not modifying RHI::SwapChain but I also took a quick look at SwapChainFrameAttachment and SwapChainFrameAttachment::GetSwapChain function and it seems like it only used by RPI::WindowContext/RPI::SwapChainPass and it doesnt really access the swapchain textures directly. Only the backend RHIs uses those texture internally as needed. Hence I was wondering how RPI will be using MDSwapChain textures directly. We dont really bind Swapchain textures to SRGs so in theory you shouldnt need to access them. The only use case I can think of is if MDSwapchain is used to present across multiple gpus. But in order to do that you will need multiple windows and code around that. Anyways lets continue this conversation when we have a use case where we think MDswapchain will be needed and how it can be implemented then.